### PR TITLE
NXP-17610 multiple files for scan xml importer

### PR DESCRIPTION
**Pull request by andreinechaev**

### CODE REVIEW

File: nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        List<DocumentModel> docs = session.query("SELECT * FROM Document WHERE dc:title='Multifile'");
+        assertEquals("we should have only one File", 1, docs.size());
+
+        DocumentModel doc = docs.get(0);
+
+        Blob mainFile = (Blob) doc.getPropertyValue("file:content");
+        assertEquals("file1.txt", mainFile.getFilename());
+
+        List<Blob> attachments = (List<Blob>) doc.getPropertyValue("files:files");
+        assertEquals(2, attachments.size());
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java]: Please check each attached blob filename and content too, to be sure.

File: nuxeo-importer-xml-parser/src/test/resources/test-ImporterMapping-Multi-files-contrib.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<component name="org.nuxeo.ecm.platform.importer.xml.parser.ImporterComponent.multifiles.contrib" version="1.0">
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/test/resources/test-ImporterMapping-Multi-files-contrib.xml]: Please use "test" somewhere in this test component name.

File: nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        session.save();
+
+        List<DocumentModel> docs = session.query("SELECT * FROM Document WHERE dc:title='Multifile'");
+        assertEquals("we should have only one File", 1, docs.size());
+
+        DocumentModel doc = docs.get(0);
+
+        Blob mainFile = (Blob) doc.getPropertyValue("file:content");
+        assertEquals("file1.txt", mainFile.getFilename());
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java]: Please also check the blob content.

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java [hunk 4]
```diff
@@ -328,8 +329,13 @@ protected void processDocAttributes(DocumentModel doc, Element el, AttributeConf
...
-                value = (Serializable) resolveComplex(el, conf);
-                if (value != null && !conf.getMapping().isEmpty()) {
+                Map<String, Object> props = (Map<String, Object>) resolveComplex(el, conf);
+                value = new HashMap<>(props);
+                if (props.containsKey("file")) {
+                    Blob blob = resolveBlob(el, conf, "file");
+                    props.put("file", blob);
+                    property.addValue(props);
+                } else {
                     property.addValue(value);
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: It seems that `property.addValue(value)` can be factored after the `if`.

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java [hunk 5]
```diff
@@ -146,7 +145,7 @@ protected DocConfigDescriptor getDocCreationConfig(Element el) {
         return null;
     }
 
-    protected List<AttributeConfigDescriptor> getAttributConfigs(Element el) {
+    protected List<AttributeConfigDescriptor> getAttributeConfigs(Element el) {
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: Please undo this typo fix or do it in a separate cleanup commit.

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java [hunk 6]
```diff
@@ -328,8 +329,13 @@ protected void processDocAttributes(DocumentModel doc, Element el, AttributeConf
                     }
                 }
             } else {
-                value = (Serializable) resolveComplex(el, conf);
-                if (value != null && !conf.getMapping().isEmpty()) {
+                Map<String, Object> props = (Map<String, Object>) resolveComplex(el, conf);
+                value = new HashMap<>(props);
+                if (props.containsKey("file")) {
+                    Blob blob = resolveBlob(el, conf, "file");
+                    props.put("file", blob);
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: Please use constants for `"file"`. Use several constants if they have different roles (I don't know from just reading the diff).

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java (hunk 4 above)
Comment (andreinechaev) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: It doesn't look so. 342 line doesn't use the value at all.

File: nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java [hunk 7]
```diff
@@ -78,9 +81,21 @@ public void test() throws Exception {
         DocumentModel doc = docs.get(0);
 
         Blob mainFile = (Blob) doc.getPropertyValue("file:content");
+        BufferedReader reader = new BufferedReader(new InputStreamReader(mainFile.getStream()));
+        String result = reader.readLine();
+        assertEquals("file1", result);
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java]: Don't reinvent the wheel, just use `Blob.getString` to get the string content for tests.

File: nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java [hunk 8]
```diff
@@ -78,9 +81,21 @@ public void test() throws Exception {
...
-        List<Blob> attachments = (List<Blob>) doc.getPropertyValue("files:files");
+        List<Map<String, Blob>> attachments = (List<Map<String, Blob>>) doc.getPropertyValue("files:files");
+        String[] filenames = new String[] {"file2.txt", "file3.txt"};
+        String[] contents = new String[] {"file2", "file3"};
+        for (int i = 0; i < filenames.length; i++) {
+            Blob blob = attachments.get(i).get("file");
+            reader = new BufferedReader(new InputStreamReader(blob.getStream()));
+            result = reader.readLine();
+            assertEquals(contents[i], result);
+            assertEquals(filenames[i], blob.getFilename());
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/test/java/org/nuxeo/ecm/platform/importer/xml/parser/test/TestMapperServiceMultiFiles.java]: Tests are more readable when they're explicit, don't do a loop:

    assertEquals(2, attachments.size());
    Blob blob2 = attachments.get(0).get("file");
    assertEquals("file2", blob2.getFilename());
    assertEquals("file2.txt", blob2.getString());
    Blob blob3 = attachments.get(1).get("file");
    assertEquals("file3", blob3.getFilename());
    assertEquals("file3.txt", blob3.getString());

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java [hunk 9]
```diff
@@ -64,6 +64,14 @@
  */
 public class XMLImporterServiceImpl {
 
+    private static final String FILE_PROPERTY = "file";
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: Don't use `private`, it prevents reuse and subclassing.

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java (hunk 4 above)
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: But `value = new HashMap<>(props)` so they're effectively the same.

File: nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java [hunk 10]
```diff
@@ -338,12 +338,13 @@ protected void processDocAttributes(DocumentModel doc, Element el, AttributeConf
                 }
             } else {
                 Map<String, Object> props = (Map<String, Object>) resolveComplex(el, conf);
-                value = new HashMap<>(props);
                 if (props.containsKey(FILE_PROPERTY)) {
                     Blob blob = resolveBlob(el, conf, FILE_PROPERTY);
                     props.put(FILE_PROPERTY, blob);
+
```
Comment (efge) [inline nuxeo-importer-xml-parser/src/main/java/org/nuxeo/ecm/platform/importer/xml/parser/XMLImporterServiceImpl.java]: Unneeded empty line.

**Comment by efge** [comment]
👍

**Comment by kevinleturc** [comment]
👍

**Pull request merged**